next merge: Helper NetworkManagerSettings.get_connections_by_id()

bernhardkaindl opened the pull request:
In PR #6, where I initially added a `connection_by_id()` function for the purpose of that example, I didn't account for the fact that NetworkManager can have many connections with the same `id`. Only the `uuid` is guaranteed to be unique.

For a helper method like you suggested to be provided by [python-sdbus-networkmanager](https://github.com/python-sdbus/python-sdbus-networkmanager), we should be able to potentially return multiple connections, so I'd return a list of them.

This PR provides this helper in the `NetworkManagerSettings` class, alongside the `get_connection_by_uuid` which is provided as part of the DBus interface by this class.

An example use in examples which add connections can look like this:
```py
    if NetworkManagerSettings().get_connections_by_id(args.conn_id):
        print(f'Connections using ID "{args.conn_id}" exist, remove them:')
        print(f'Run: nmcli connection delete "{args.conn_id}"')
        return
```

bernhardkaindl commented:
@igo95862: Rebased to the merge commit of #7, this could be merged next!

igo95862 commented on sdbus_async/networkmanager/objects.py:
```diff
@@ -154,6 +154,21 @@ def __init__(self, bus: Optional[SdBus] = None) -> None:
+        for connection_path in connection_paths:
+            settings = NetworkConnectionSettings(connection_path)
+            settings_properites = await settings.get_settings()
+            if settings_properites["connection"]["id"][1] == connection_id:
```
Are you sure that `[1]` is needed. 

This is a part of config what I have in my NetworkManager `GetSettings()` method:

```
 'connection': {'id': 'Ethernet connection 1',           
```

That means `["connection"]["id"]` would indeed get the pretty name of the connection but `[1]` would get the second character of the string.

I looked at upstream documentation and it says the id can only be a string: https://developer-old.gnome.org/NetworkManager/unstable/nm-settings-dbus.html (no arrays possible)

bernhardkaindl replied:
I know that the nested dict returned by NetworkConnectionSettings(connection_path).get_settings() is confusing, and I even developed a helper namespace to abstract this confusion away, which I would propose to be merged as well, then the code would be less confusing.

Your (and likely every readers) first guess (and mine was of course too) is that ["connection"]["id"] is a string as documented, but actually every setting in [nm-settings-dbus.html](https://developer-old.gnome.org/NetworkManager/unstable/nm-settings-dbus.html) is not the actual data, but aways a tuple of ("datatype", "value):

If you'd like to see yourself, you'd get this if you'd a debug print:
``` 
print(settings_properites["connection"]["id"])
("s", "Ethernet connection 1)
```

Because the nm setting `connection.id` is a string, in the nested dict, `["connection"]["id"][0]` (the 1st entry of the tuple) is actually `"s"` (I think this is called the "signature") which says that the value, which is the second entry of the tuple can be used a string, so `["connection"]["id"][1]` is e.g. `"Ethernet connection 1"`, the value of this dbus setting.

In some other draft for an example I even created a small function to move the [1] out of sight, the smalles version would be:
```py 
def SettingsValue(setting) {
    return setting[1]
}
```
But of couse it makes the code longer, not shorter. Thus, what I worked on to solve this would be some a SimpleNamespace or DotDict. But to keep this as simple as possible, I just put the [1] here. Maybe the next best solution would be to use a tiny helper function like the one above or this:
```
Value = 1
...
if settings_properites["connection"]["id"][Value] == connection_id:
```
I guess this latter solution is most transparent for a reader, I guess I should use it for now until the nested SimpleNamespace I created is submitted and merged. Then this could be changed to:
```py
if settings_properites.connection.id == connection_id:
```

igo95862 replied:
You are correct.

I totally forgot how that Variant type will return the type signature followed by type itself.

I never really liked how I implemented the Variant type in python-sdbus. I might think of better way of implementing it but must be backwards compatible.

igo95862 commented on sdbus_async/networkmanager/objects.py:
```diff
@@ -154,6 +154,21 @@ def __init__(self, bus: Optional[SdBus] = None) -> None:
 
+    async def get_connections_by_id(self, connection_id: str) -> List[str]:
+        """Helper method to get a list of connection profile paths
+        which use the given connection identifier.
```
Can you say that it uses the pretty name and give an example (for example `Ethernet connection 1`)?

bernhardkaindl replied:
@igo95862 Added to the other change, updated commit is pushed now, ready to merge.

bernhardkaindl commented:
This is the new blocking version with the improved docstring and the "annotated" version of reading the value of `id`:
```py
    def get_connections_by_id(self, connection_id: str) -> List[str]:
        """Helper method to get a list of connection profile paths
        which use the given connection identifier.

        :param str connection_id: The connection identifier of the connections,
        e.g. "Ethernet connection 1"
        :return: List of connection profile paths using the given identifier.
        """
        connection_paths_with_matching_id = []
        value = 1
        for connection_path in self.connections:
            s = NetworkConnectionSettings(connection_path)
            if s.get_settings()["connection"]["id"][value] == connection_id:
                connection_paths_with_matching_id.append(connection_path)
        return connection_paths_with_matching_id
```
Th asyncio version is changed likewise. I think this addresses your remarks so far.

igo95862 commented:
`["connection"]["id"][1]` was a better option.

It is my mistake that I forgot about the variant type.

igo95862 commented:
I think I will extend the tuple class that Variant returns. Add something like unwrap as string and etc...

igo95862 commented:
I think the entire settings dictionary can be replicated with some [dataclass](https://docs.python.org/3/library/dataclasses.html). This would allow to document all settings that the objects can have and allow typing the settings. Every method that returns or takes a settings dictionary would have a sibling helper method that would return the dataclass.

bernhardkaindl commented:
> `["connection"]["id"][1]` was a better option.

pushed.

bernhardkaindl commented:
Thanks for the approval. Since I'm not invited as a developer, of course only you can do the merge.

BTW, in these discussions, there are some nice approaches on creating classes which could be a base for providing a class-namespace on top of nested dicts with lists and tuples:

https://stackoverflow.com/questions/13520421/recursive-dotdict
https://stackoverflow.com/questions/3031219/recursively-access-dict-via-attributes-as-well-as-index-access

I think one of the examples (I hope I can find it again) which I saw even said that writing to the class writes even to the underlying dict.

That feature (writing to the underlying dict on changes, ) would be essential for using such dict in an `update_connection` example.

igo95862 commented:
> BTW, in these discussions, there are some nice approaches on creating classes which could be a base for providing a class-namespace on top of nested dicts with lists and tuples:

I would go with dataclass from standard library. The typing information can be implemented with it.

Since there is upstream documentation on all values for all settings dictionaries I would implement them tomorrow.

Thank you again for working on pull requests.
